api/src/main/java/net/kyori/adventure/pointer/Pointers.java

```diff
@@ -0,0 +1,159 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.pointer;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+import net.kyori.adventure.util.Buildable;
+import net.kyori.adventure.util.TriState;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
+
+/**
+ * A collection of {@link Pointer pointers}.
+ *
+ * @since 4.8.0
+ */
+public interface Pointers extends Buildable<Pointers, Pointers.Builder> {
+  /**
+   * An empty set of pointers.
+   *
+   * @since 4.8.0
+   */
+  @NotNull Pointers EMPTY = PointersImpl.EMPTY;
+
+  /**
+   * Gets a new pointers builder.
+   *
+   * @return the builder
+   * @see Builder
+   * @since 4.8.0
+   */
+  @Contract(pure = true)
+  static @NotNull Builder builder() {
+    return new PointersImpl.BuilderImpl();
+  }
+
+  /**
+   * Gets the value of {@code pointer}.
+   *
+   * @param pointer the pointer
+   * @param <T> the type
+   * @return the value
+   * @since 4.8.0
+   */
+  <T> @NotNull Optional<T> get(final @NotNull Pointer<T> pointer);
+
+  /**
+   * Gets the value of {@code pointer}.
+   *
+   * <p>If this {@code Audience} is unable to provide a value for {@code pointer}, {@code defaultValue} will be returned.</p>
+   *
+   * @param pointer the pointer
+   * @param defaultValue the default value
+   * @param <T> the type
+   * @return the value
+   * @since 4.8.0
+   */
+  @Contract("_, !null -> !null; _, null -> null")
+  @SuppressWarnings("checkstyle:MethodName")
+  default <T> @Nullable T getOrDefault(final @NotNull Pointer<T> pointer, final @Nullable T defaultValue) {
+    return this.get(pointer).orElse(defaultValue);
+  }
+
+  /**
+   * Gets the value of {@code pointer}.
+   *
+   * <p>If this {@code Audience} is unable to provide a value for {@code pointer}, the value supplied by {@code defaultValue} will be returned.</p>
+   *
+   * @param pointer the pointer
+   * @param defaultValue the default value supplier
+   * @param <T> the type
+   * @return the value
+   * @since 4.8.0
+   */
+  @SuppressWarnings("checkstyle:MethodName")
+  default <T> @UnknownNullability T getOrDefaultFrom(final @NotNull Pointer<T> pointer, final @NotNull Supplier<? extends T> defaultValue) {
+    return this.get(pointer).orElseGet(defaultValue);
+  }
+
+  /**
+   * Checks if these pointers contain a value for the given pointer.
+   *
+   * @param pointer the pointer
+   * @param <T> the type
+   * @return a tri-state
+   * @since 4.8.0
+   */
+  <T> @NotNull TriState has(final @NotNull Pointer<T> pointer);
```

Comment:
Not sure if it makes sense for this to be a `TriState`.

From looking at the implementation, it seems that this returns NOT_SET if the pointer is not known to this `Pointers` instance, and TRUE/FALSE based on the nullness of the value, but I'm not sure that distinction is useful.

Reply:
Perhaps a different name for this method? `supports`?

Reply:
A different enum could work, `Pointer.State` with `PRESENT`, `ABSENT` and `NOT_SUPPORTED` perhaps?

Reply:
Not a fan of a new enum for this.

Reply:
I think a boolean `supports()` like kashike suggested would make the most sense here, just indicating whether a Pointer is supported or not, and saying nothing about its value.

Reply:
Ah I like that! So it essentially just returns the result of `containsKey()`?

Reply:
Yea